### PR TITLE
Support filter push down in Big Query

### PR DESCRIPTION
**Pull request by davidrabinowitz**
Added support for filter push down. This has two benefits of this are:

- Performance improvement, where row filtering is done on the BigQuery storage size where applicable
- Partitioned tables where filtering by the partition is mandatory are supported now. 

In addition, the BigQuery libraries were updated.

### CODE REVIEW

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                        case BELOW:
+                            rangeConjuncts.add(toPredicate(columnName, "<", range.getHigh().getValue(), column));
+                            break;
+                        default:
+                            throw new AssertionError("Unhandled bound: " + range.getHigh().getBound());
+                    }
+                }
+                // If rangeConjuncts is null, then the range was ALL, which should already have been checked for
+                checkState(!rangeConjuncts.isEmpty());
+                disjuncts.add("(" + Joiner.on(" AND ").join(rangeConjuncts) + ")");
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:135]: Use `concat()` to join the conjuncts?

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        return domains.map(this::toConjuncts)
+                .map(this::concat)
+                .map(filter -> filter
+                        .replace(_PARTITIONDATE.getName(), _PARTITIONDATE.getName().toUpperCase())
+                        .replace(_PARTITIONTIME.getName(), _PARTITIONTIME.getName().toUpperCase()));
+    }
+
+    private String concat(List<String> clauses)
+    {
+        return clauses.isEmpty() ? null : clauses.stream().collect(joining(" AND "));
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:60]: Instead of 

```java
clauses.stream().collect(joining(" AND "))
```
do 

```java
String.join(" AND ", clauses)
```

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            }
+        }
+
+        // Add back all of the possible single values either as an equality or an IN predicate
+        if (singleValues.size() == 1) {
+            disjuncts.add(toPredicate(columnName, "=", getOnlyElement(singleValues), column));
+        }
+        else if (singleValues.size() > 1) {
+            String values = singleValues.stream()
+                    .map(value -> column.getBigQueryType().convertToString(value))
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:145]:
```suggestion
                    .map(column.getBigQueryType()::convertToString)
```

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            disjuncts.add(quote(columnName) + " IN (" + values + ")");
+        }
+
+        // Add nullability disjuncts
+        checkState(!disjuncts.isEmpty());
+        if (domain.isNullAllowed()) {
+            disjuncts.add(quote(columnName) + " IS NULL");
+        }
+
+        return disjuncts.stream().collect(joining(" OR ", "(", ")"));
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:156]: Maybe:

```suggestion
        return "(" + String.join(" OR ", disjuncts) + ")";
```

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryMetadata.java [hunk 5]
```diff
@@ -138,6 +146,7 @@ private TableInfo getBigQueryTable(SchemaTableName tableName)
 
     public ConnectorTableMetadata getTableMetadata(ConnectorSession session, SchemaTableName schemaTableName)
     {
+        log.debug("getTableMetadata(session=%s, schemaTableName=%s)", session, schemaTableName);
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryMetadata.java:149]: Remove this

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryMetadata.java [hunk 6]
```diff
@@ -151,24 +160,57 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
...
+                if (timePartitioning.getField() != null) {
+                    properties.put("time_partitioning", timePartitioning.getField());
+                }
+                else {
+                    properties.put("time_partitioning", _PARTITIONTIME.getName());
+                }
+            }
+        }
+        Optional<String> comment = Optional.ofNullable(table.getDescription());
+        return new ConnectorTableMetadata(schemaTableName, columns, properties.build(), comment);
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryMetadata.java:187]: Is this related to filter pushdown? If not, pull it out into a separate commit.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadSessionCreator.java [hunk 7]
```diff
@@ -73,9 +75,11 @@ public ReadSessionCreator(
 
         TableInfo actualTable = getActualTable(tableDetails, selectedFields, new String[] {});
 
+        List<String> filteredSelectedFields = selectedFields.stream().filter(BigQueryUtil::validColumnName).collect(toList());
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadSessionCreator.java:78]: Formatting:

```suggestion
        List<String> filteredSelectedFields = selectedFields.stream()
                .filter(BigQueryUtil::validColumnName)
                .collect(toList());
```

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.plugin.bigquery.BigQueryColumnHandle._PARTITIONDATE;
+import static io.prestosql.plugin.bigquery.BigQueryColumnHandle._PARTITIONTIME;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+
+class BigQueryFilterQueryBuilder
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:37]: Having an helper object actually helps the implementation in my perspective, but I've changed the interface

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+{
+    private static final String QUOTE = "`";
+    private static final String ESCAPED_QUOTE = "``";
+    private TupleDomain<ColumnHandle> tupleDomain;
+
+    BigQueryFilterQueryBuilder(TupleDomain<ColumnHandle> tupleDomain)
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:43]: This class is package protected. I any case, the constructor was changed to `private`

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        this.tupleDomain = tupleDomain;
+    }
+
+    Optional<String> buildFilter()
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:48]: Even for package protected class?

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    {
+        Optional<Map<ColumnHandle, Domain>> domains = tupleDomain.getDomains();
+        return domains.map(this::toConjuncts)
+                .map(this::concat)
+                .map(filter -> filter
+                        .replace(_PARTITIONDATE.getName(), _PARTITIONDATE.getName().toUpperCase())
+                        .replace(_PARTITIONTIME.getName(), _PARTITIONTIME.getName().toUpperCase()));
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:55]: I had issues with those two fields, as they are pseudo columns of insert time partitioned tables, and the fact that Presto change all the column names to lower case.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:60]: Can you please have a second look? `Optional.map()` returns `Optional.empty()` if the mapper function returns a null value, not `Optional.of(null)`.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (martint) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:43]: Yes. Please see https://github.com/prestosql/presto/pull/3302#discussion_r402458480

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (martint) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:60]: Yes, you're right!

**Comment by davidrabinowitz** [comment]
Hi @martint , all your comments have been resolved.

**Comment by martint** [comment]
Can you squash the commits that address review comments into the corresponding commit? Also, change the commit message of the first one to "Support filter pushdown" (use the imperative mood. BTW, here's a good article that describes how we try to formulate commit messages: https://chris.beams.io/posts/git-commit/)

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java [hunk 12]
```diff
@@ -73,4 +49,59 @@ public ReadRowsHelper(BigQueryStorageClient client, ReadRowsRequest.Builder requ
...
+        }
+
+        @Override
+        public ReadRowsResponse next()
+        {
+            do {
+                try {
+                    ReadRowsResponse response = serverResponses.next();
+                    readRowsCount += response.getRowCount();
+                    //logDebug(s"read ${response.getSerializedSize} bytes");
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:86]: Remove commented out code

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java [hunk 13]
```diff
@@ -73,4 +49,59 @@ public ReadRowsHelper(BigQueryStorageClient client, ReadRowsRequest.Builder requ
                 .call(readRowsRequest.build())
                 .iterator();
     }
+
+    // Ported from https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/150
+    static class ReadRowsIterator
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:54]: This can be private

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java [hunk 14]
```diff
@@ -73,4 +49,59 @@ public ReadRowsHelper(BigQueryStorageClient client, ReadRowsRequest.Builder requ
...
+            implements Iterator<ReadRowsResponse>
+    {
+        ReadRowsHelper helper;
+        Storage.StreamPosition.Builder readPosition;
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:58]: These can be `private final`

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java [hunk 15]
```diff
@@ -73,4 +49,59 @@ public ReadRowsHelper(BigQueryStorageClient client, ReadRowsRequest.Builder requ
...
+        Iterator<ReadRowsResponse> serverResponses;
+        long readRowsCount;
+        int retries;
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:61]: These can be private

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java [hunk 16]
```diff
@@ -14,12 +14,12 @@
 package io.prestosql.plugin.bigquery;
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:14]: Change the commit message to "Improve readRows performance"

Also, explain in the commit message improves performance. I see a refactoring, but I can't figure out what's the difference between the two versions that will improve performance.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryConnectorModule.java [hunk 17]
```diff
@@ -46,6 +46,22 @@ public static HeaderProvider createHeaderProvider(NodeManager nodeManager)
         return FixedHeaderProvider.create("user-agent", "prestosql/" + nodeManager.getCurrentNode().getVersion());
     }
 
+    // Note that at this point the config has been validated, which means that option 2 or option 3 will always be valid
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryConnectorModule.java:49]: Is this resolved?

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java [hunk 18]
```diff
@@ -33,25 +33,53 @@
...
     {
         this.name = requireNonNull(name, "column name cannot be null");
         this.bigQueryType = requireNonNull(bigQueryType, () -> format("column type cannot be null for column [%s]", name));
         this.mode = requireNonNull(mode, "Field mode cannot be null");
         this.subColumns = ImmutableList.copyOf(requireNonNull(subColumns, "subColumns is null"));
         this.description = description;
+        this.hidden = hidden;
+    }
+
+    BigQueryColumnHandle(
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java:75]: Annotate this with `@VisibleForTesting`

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    private Optional<String> buildFilter()
+    {
+        Optional<Map<ColumnHandle, Domain>> domains = tupleDomain.getDomains();
+        return domains.map(this::toConjuncts)
+                .map(this::concat)
+                .map(filter -> filter
+                        .replace(PARTITIONDATE.getName(), PARTITIONDATE.getName())
+                        .replace(PARTITIONTIME.getName(), PARTITIONTIME.getName()));
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryFilterQueryBuilder.java:57]: I still don't understand how this is anything other than a no-op in its current form.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryType.java [hunk 20]
```diff
@@ -133,6 +148,77 @@ static long toPrestoTimestamp(String datetime)
...
+        return Decimals.toString(slice, NUMERIC_DATA_TYPE_SCALE);
+    }
+
+    static String bytesToStringConverter(Object value)
+    {
+        Slice slice = (Slice) value;
+        return format("FROM_BASE64('%s')", Base64.getEncoder().encodeToString(slice.getBytes()));
+    }
+
+    private static String quote(String s)
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryType.java:212]: Don't abbreviate variable name. Rename it to `value`

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java [hunk 21]
```diff
@@ -51,4 +56,14 @@ static BigQueryException convertToBigQueryException(BigQueryError error)
     {
         return new BigQueryException(UNKNOWN_CODE, error.getMessage(), error);
     }
+
+    static <T> Function<T, T> identity()
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java:60]: This is not used

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java [hunk 22]
```diff
@@ -29,6 +31,9 @@
             "Connection closed with unknown cause",
             "Received unexpected EOS on DATA frame from server");
 
+    static final ImmutableSet<String> INVALID_COLUMN_NAMES = ImmutableSet.<String>builder()
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java:34]: Make it `private`. Also, use `Set` as the declared type.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java [hunk 23]
```diff
@@ -51,4 +56,14 @@ static BigQueryException convertToBigQueryException(BigQueryError error)
     {
         return new BigQueryException(UNKNOWN_CODE, error.getMessage(), error);
     }
+
+    static <T> Function<T, T> identity()
+    {
+        return t -> t;
+    }
+
+    static boolean validColumnName(String columnName)
```
Comment (martint) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java:65]: `public`

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java (hunk 16 above)
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:14]: @martint This change is a port from the spark connector - instead of reading the entire stream as Avro and them converting it to Presto pages, the read from the server is done in a lazy fashion using the new Iterator. We have seen performance improvement especially when reading part of the data (for example when using `LIMIT`)

Comment (martint) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:14]: Got it. Please add that explanation to the commit message

Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java:14]: Just did...

**Comment by davidrabinowitz** [comment]
Thanks @martint !

**Comment by martint** [comment]
There's a build failure. Oddly, I wasn't able to reproduce it locally. Are there any elements to it that would make it non-deterministic?

```
[ERROR] testNoFailures(io.prestosql.plugin.bigquery.TestReadRowsHelper)  Time elapsed: 0.001 s  <<< FAILURE!
java.lang.NullPointerException
	at com.google.protobuf.SingleFieldBuilderV3.getBuilder(SingleFieldBuilderV3.java:133)
	at com.google.cloud.bigquery.storage.v1beta1.Storage$ReadRowsRequest$Builder.getReadPositionBuilder(Storage.java:7995)
	at io.prestosql.plugin.bigquery.ReadRowsHelper.readRows(ReadRowsHelper.java:42)
	at io.prestosql.plugin.bigquery.TestReadRowsHelper.testNoFailures(TestReadRowsHelper.java:50)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by findepi** [comment]
@martint `ReadRowsRequest.Builder request` is used by test threads.
Make it test-local or mark class `@Test(singleThreaded = true)`

**Comment by findepi** [comment]
https://github.com/prestosql/presto/pull/3504

**Comment by martint** [comment]
Merged. Thanks for your contribution!

**Comment by tooptoop4** [comment]
@davidrabinowitz any plan to add aggregate pushdown like in https://github.com/prestosql/presto/commit/193872a247c49099c6eeda4640f65e36222a4ad0#diff-e4b92558819e35928d8d20100000291dda1e15cf907c1d2f06ba4015b0afd6cc ?
